scala-generator/src/main/scala/models/http4s/server/Route.scala

```diff
@@ -57,16 +68,30 @@ case class Route(ssd: ScalaService, resource: ScalaResource, op: ScalaOperation,
 
     val params =
       Seq(Some(s"_req: ${config.requestClass}")) ++
-      op.nonHeaderParameters.map { field =>
-        val typ = field.datatype match {
-          case ScalaDatatype.Option(container: ScalaDatatype.Container) => container.name
-          case other => other.name
-        }
-        Some(s"${ScalaUtil.quoteNameIfKeyword(field.name)}: $typ")
-      } ++
-      Seq(op.body.map(body => s"body: => ${config.generateDecodeResult(body.datatype.name)}"))
+        nonHeaderParameters ++
+        Seq(op.body.map(body => s"body: => ${config.generateDecodeResult(body.datatype.name)}"))
+
+    val requestCaseClass = s"$requestCaseClassName(${nonHeaderParameters.flatten.mkString(", ")})"
+
+    val requestDecoder = Seq(s"implicit val ${requestCaseClassName}Decoder: _root_.io.circe.Decoder[$requestCaseClassName] = _root_.io.circe.Decoder.instance { a =>",
+                             s"  for {",
+                             op.nonHeaderParameters.map(p => s"""    ${p.name} <- a.downField("${p.originalName}").as[${p.datatype.name}]""").mkString("\n"),
+                             s"  } yield {",
+                             s"    $requestCaseClassName(",
+                             op.nonHeaderParameters.map(p => s"""      ${p.name} = ${p.name}""").mkString(",\n"),
+                             s"    )",
+                             s"  }",
+                             s"}")
+
+    val requestCaseClassAndDecoder = op.method match {
+      case Method.Get | Method.Delete => Seq()
+      case _ => Seq(
+        s"case class $requestCaseClass",
+        s"") ++ requestDecoder ++ Seq("")
```

Comment:
Do we need to generate the request case class for all Post and Put? I would think we only need it if the operation has form params, no?

Reply:
Good call, I'll actually use the same logic to decide whether to generate the contentType check below or not